File: src/simulation/GBObject.java
/*******************************************************************************
 * Copyright (c) 2002-2013 (c) Devon and Warren Schudy
 * Copyright (c) 2014  Devon and Warren Schudy, Mike Anderson
 *******************************************************************************/
package simulation;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.RenderingHints;
import java.awt.geom.Ellipse2D;
import java.awt.image.BufferedImage;

import Rendering.GBProjection;
import sides.Side;
//Maps FinePoints to screen locations
import support.FinePoint;

public abstract class GBObject {
	FinePoint position;
	FinePoint velocity;

	public BufferedImage image;

	protected double radius;
	protected double mass;

	public static final double kCollisionRepulsion = 0.02;
	public static final double kCollisionRepulsionSpeed = 0.1;
	public static final double kMaxOverlap = 0.04;
	public static final short kMaxSquareMiniSize = 4;

	// Some abstract functions aren't implemented by all descendants.
	// Java hates this so they were made protected and given default behavior
	// here.
	protected void takeDamage(double amount, Side origin) {
	}

	protected double takeEnergy(double amount) {
		return 0;
	}

	protected double giveEnergy(double amount) {
		return 0;
	}

	protected double maxTakeEnergy() {
		return 0;
	}

	protected double maxGiveEnergy() {
		return 0;
	}

	// high-level actions
	protected void think(GBWorld world) {
	} // Not everything has a brain

	public abstract void act(GBWorld world); // Everything can act

	protected void collideWithWall() {
	}

	protected void collideWith(GBObject other) {
	}

	protected void collectStatistics(ScoreKeeper keeper) {
	}

	// high-level queries
	public abstract GBObjectClass getObjectClass();

	public Side getOwner() {
		return null;
	} // food is not owned by anyone

	protected double getEnergy() {
		return 0;
	}

	protected double getInterest() {
		return 0;
	} // how attractive to autocamera
	
	public String getDetails() {
		return "";
	}

	// evil antimodular drawing code
	public abstract Color getColor();

	public abstract void draw(Graphics g, GBProjection proj,
			boolean detailed);

	public void drawUnderlay(Graphics g, GBProjection proj,
			boolean detailed) {
	}

	public void drawOverlay(Graphics g, GBProjection proj,
			boolean detailed) {
	}

	public GBObject next;// references the next object; allows GBObject to act
							// as a singly-linked list

	public GBObject(FinePoint where, double r) {
		position = where;
		radius = r;
		velocity = new FinePoint();
		mass = 0;
		next = null;
	}

	public GBObject(FinePoint where, double r, FinePoint vel) {
		position = where;
		radius = r;
		velocity = vel;
		mass = 0;
		next = null;
	}

	public GBObject(FinePoint where, double r, double m) {
		position = where;
		radius = r;
		velocity = new FinePoint();
		mass = m;
		next = null;
	}

	public GBObject(FinePoint where, double r, FinePoint vel, double m) {
		position = where;
		radius = r;
		velocity = vel;
		mass = m;
		next = null;
	}

	public FinePoint getPosition() {
		return position;
	}

	public void setPosition(FinePoint newPosition) {
		position = newPosition;
	}

	public void moveBy(FinePoint delta) {
		position = position.add(delta);
	}

	public void moveBy(double deltax, double deltay) {
		position.x += deltax;
		position.y += deltay;
	}

	public double getLeft() {
		return position.x - radius;
	}

	public double getTop() {
		return position.y + radius;
	}

	public double getRight() {
		return position.x + radius;
	}

	public double getBottom() {
		return position.y - radius;
	}

	public FinePoint getVelocity() {
		return velocity;
	}

	public double getSpeed() {
		return velocity.norm();
	}

	public void setVelocity(double sx, double sy) {
		velocity.set(sx, sy);
	}

	public void setSpeed(double speed) {
		velocity.setNorm(speed);
	}

	public void accelerate(FinePoint deltav) {
		velocity = velocity.add(deltav);
	}

	public void drag(double friction, double linearCoeff, double quadraticCoeff) {
		if (velocity.isZero())
			return;
		double speed = getSpeed();
		speed -= speed * (speed * quadraticCoeff + linearCoeff);
		if (speed > friction)
			setSpeed(speed - friction);
		else
			setVelocity(0, 0);
	}

	public double getRadius() {
		return radius;
	}

	public double getMass() {
		return mass;
	}

	public boolean intersects(GBObject other) {
		double r = radius + other.radius;
		return other.position.x - position.x < r
				&& position.x - other.position.x < r
				&& other.position.y - position.y < r
				&& position.y - other.position.y < r
				&& position.inRange(other.position, r);
	}

	public double overlapFraction(GBObject other) {
		if (radius == 0)
			return 0;
		// returns the fraction of our radius that other overlaps
		double dist = (position.subtract(other.position)).norm(); // center-to-center
																	// dist
		dist = Math.max(radius + other.radius - dist, 0); // overlap of edges
		return Math.min(dist / radius, 1.0); // don't overlap more than 1
	}

	public void doBasicCollide(GBObject other) {
		if (intersects(other)) {
			collideWith(other);
			other.collideWith(this);
		}
	}

	public void doSolidCollide(GBObject other, double coefficient) {
		if (intersects(other)) {
			collideWith(other);
			other.collideWith(this);
			doElasticBounce(other, coefficient);
		}
	}

	public void pushBy(FinePoint impulse) {
		if (mass != 0)
			velocity = velocity.add(impulse.divide(mass));
	}

	public void pushBy(double impulse, double dir) {
		pushBy(FinePoint.makePolar(impulse, dir));
	}

	public void push(GBObject other, double impulse) {
		FinePoint cc = other.position.subtract(position); // center-to-center
		if (cc.normSquare() == 0)
			return; // don't push if in same location
		other.pushBy(cc.divide(cc.norm()).multiply(impulse));
	}

	public void doElasticBounce(GBObject other, double coefficient /*
																 * TODO:
																 * substitute
																 * default value
																 * if needed: =1
																 */) {
		double m1 = mass;
		double m2 = other.mass;
		if (m1 == 0 || m2 == 0)
			return; // can't bounce massless objects
		FinePoint cc = other.position.subtract(position); // center-to-center
		if (cc.normSquare() == 0)
			cc.set(1, 0); // if in same place, pick an arbitrary cc
		double dist = cc.norm();
		FinePoint rv1 = velocity.projection(cc); // radial component of velocity
		FinePoint rv2 = other.velocity.projection(cc);
		if (velocity.dotProduct(cc) - other.velocity.dotProduct(cc) > 0) {
			// if still moving closer...
			FinePoint center = (rv1.multiply(m1).add(rv2.multiply(m2)))
					.divide((m1 + m2)); // velocity of center-of-mass
			accelerate(rv2.subtract(center).multiply(m2 * coefficient)
					.divide(m1).subtract(rv1.subtract(center)));
			other.accelerate(rv1.subtract(center).multiply(m1 * coefficient)
					.divide(m2).subtract(rv2.subtract(center)));
		}
		double totalr = radius + other.radius;
		double overlap = totalr - dist;
		if (overlap > kMaxOverlap) {
			FinePoint away1 = cc.negate().divide(dist).multiply(m2)
					.divide(m1 + m2);
			FinePoint away2 = cc.divide(dist).multiply(m1).divide(m1 + m2);
			moveBy(away1.multiply(kCollisionRepulsionSpeed));
			other.moveBy(away2.multiply(kCollisionRepulsionSpeed));
			accelerate(away1.multiply(kCollisionRepulsion));
			other.accelerate(away2.multiply(kCollisionRepulsion));
		}
	}

	public void move() {
		position = position.add(velocity);
	}

	/**
	 * Returns a rectangle (approximately) representing the position of the
	 * rendered object on the screen.
	 * 
	 * @param proj
	 * @return
	 */
	protected Rectangle getScreenRect(GBProjection proj) {
		int oWidth = (int) Math.max(radius * proj.getScale() * 2, 1);
		return new Rectangle(proj.toScreenX(getLeft()), proj.toScreenY(getTop()), oWidth, oWidth);
	}
	
	protected Ellipse2D.Double getScreenEllipse(GBProjection proj) {
		double oWidth = Math.max(radius * proj.getScale() * 2, 1);
		return new Ellipse2D.Double(proj.toScreenX(getLeft()), proj.toScreenY(getTop()), oWidth, oWidth);
	}

	public void drawShadow(Graphics g, GBProjection proj,
			FinePoint offset, Color color) {
		Graphics2D g2d = (Graphics2D) g;
		Ellipse2D.Double where = getScreenEllipse(proj);
		int scale = proj.getScale();
		where.x += offset.x * scale;
		where.y -= offset.y * scale;
		g2d.setPaint(color);
		g2d.fill(where);
	}

	protected void drawImage(Graphics g, GBProjection proj) {
		if (image == null)
			return;
		int x1 = proj.toScreenX(getLeft());
		int y1 = proj.toScreenY(getTop());
		int x2 = proj.toScreenX(getRight());
		int y2 = proj.toScreenY(getBottom());
		if (x2 - x1 < 2)
			x2 = x1 + 2;
		if (y2 - y1 < 2)
			y2 = y1 + 2;
		((Graphics2D)g).setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BILINEAR);
		g.drawImage(image, x1, y1, x2, y2, 0, 0, image.getWidth(),
				image.getHeight(), null);
	}
};

File: grobots_cpp/simulation/GBObject.cpp
// GBObject.cpp
// Grobots (c) 2002-2004 Devon and Warren Schudy
// Distributed under the GNU General Public License.

#include "GBObject.h"
#include "GBErrors.h"
#include "GBColor.h"
#include "GBDeletionReporter.h"

const GBAccelerationScalar kCollisionRepulsion = 0.02;
const GBSpeed kCollisionRepulsionSpeed = 0.1;
const GBDistance kMaxOverlap = 0.04;

const short kMaxSquareMiniSize = 4;


GBObject::GBObject() {
	throw GBBadConstructorError();
}

GBObject::GBObject(const GBPosition & where, const GBDistance r)
	: position(where), radius(r), velocity(), mass(0),
	next(nil)
{}

GBObject::GBObject(const GBPosition & where, const GBDistance r, const GBVelocity & vel)
	: position(where), radius(r), velocity(vel), mass(0),
	next(nil)
{}

GBObject::GBObject(const GBPosition & where, const GBDistance r, const GBMass m)
	: position(where), radius(r), velocity(), mass(m),
	next(nil)
{}

GBObject::GBObject(const GBPosition & where, const GBDistance r, const GBVelocity & vel, const GBMass m)
	: position(where), radius(r), velocity(vel), mass(m),
	next(nil)
{}

GBObject::~GBObject() {}

GBPosition GBObject::Position() const {
	return position;
}

void GBObject::SetPosition(const GBPosition & newPosition) {
	position = newPosition;
}

void GBObject::MoveBy(const GBPosition & delta) {
	position += delta;
}

void GBObject::MoveBy(const GBDistance deltax, const GBDistance deltay) {
	position.x += deltax;
	position.y += deltay;
}

GBCoordinate GBObject::Left() const {
	return position.x - radius;
}

GBCoordinate GBObject::Top() const {
	return position.y + radius;
}

GBCoordinate GBObject::Right() const {
	return position.x + radius;
}

GBCoordinate GBObject::Bottom() const {
	return position.y - radius;
}

GBVelocity GBObject::Velocity() const {
	return velocity;
}

GBSpeed GBObject::Speed() const {
	return velocity.Norm();
}

void GBObject::SetVelocity(const GBSpeed sx, const GBSpeed sy) {
	velocity.Set(sx, sy);
}

void GBObject::SetSpeed(const GBSpeed speed) {
	velocity.SetNorm(speed);
}

void GBObject::Accelerate(const GBVelocity & deltav) {
	velocity += deltav;
}

void GBObject::Drag(const GBAccelerationScalar friction, const GBRatio linearCoeff, const GBRatio quadraticCoeff) {
	if ( velocity.Zero() ) return;
	GBSpeed speed = Speed();
	speed -= speed * (speed * quadraticCoeff + linearCoeff);
	if ( speed > friction )
		SetSpeed(speed - friction);
	else
		SetVelocity(0, 0);
}

GBDistance GBObject::Radius() const {
	return radius;
}

GBMass GBObject::Mass() const {
	return mass;
}

bool GBObject::Intersects(const GBObject * other) const {
	GBDistance r = radius + other->radius;
	return other->position.x - position.x < r && position.x - other->position.x < r
		&& other->position.y - position.y < r && position.y - other->position.y < r
		&& position.InRange(other->position, r);
}

GBNumber GBObject::OverlapFraction(const GBObject * other) const {
	if ( ! radius ) return 0;
// returns the fraction of our radius that other overlaps
	GBDistance dist = (position - other->position).Norm();	// center-to-center dist
	dist = max(radius + other->radius - dist, 0);	// overlap of edges
	return min(dist / radius, GBNumber(1));	// don't overlap more than 1
}

void GBObject::BasicCollide(GBObject * other) {
	if ( Intersects(other) ) {
		CollideWith(other);
		other->CollideWith(this);
	}
}

void GBObject::SolidCollide(GBObject * other, GBRatio coefficient) {
	if ( Intersects(other) ) {
		CollideWith(other);
		other->CollideWith(this);
		ElasticBounce(other, coefficient);
	}
}

void GBObject::PushBy(const GBMomentum & impulse) {
	if ( mass )
		velocity += impulse / mass;
}

void GBObject::PushBy(const GBMomentumScalar impulse, const GBAngle dir) {
	PushBy(GBFinePoint::MakePolar(impulse, dir));
}

void GBObject::Push(GBObject * other, const GBMomentumScalar impulse) const {
	GBVector cc = other->position - position; // center-to-center
	if ( cc.Norm() == 0 )
		return; // don't push if in same location
	other->PushBy(cc / cc.Norm() * impulse);
}

void GBObject::TakeDamage(const GBDamage, GBSide *) {
}

GBEnergy GBObject::TakeEnergy(const GBEnergy) {
	return 0;
}

GBEnergy GBObject::GiveEnergy(const GBEnergy) {
	return 0;
}

GBEnergy GBObject::MaxTakeEnergy() {
	return 0;
}

GBEnergy GBObject::MaxGiveEnergy() {
	return 0;
}

void GBObject::ElasticBounce(GBObject * other, GBRatio coefficient) {
	GBMass m1 = mass;
	GBMass m2 = other->mass;
	if ( ! m1 || ! m2 )
		return; // can't bounce massless objects
	GBVector cc = other->position - position; // center-to-center
	if ( ! cc.NormSquare() )
		cc.Set(1, 0); // if in same place, pick an arbitrary cc
	GBDistance dist = cc.Norm();
	GBVelocity rv1 = velocity.Projection(cc); // radial component of velocity
	GBVelocity rv2 = other->velocity.Projection(cc);
	if ( velocity.DotProduct(cc) - other->velocity.DotProduct(cc) > 0 ) {
		// if still moving closer...
		GBVelocity center = (rv1 * m1 + rv2 * m2) / (m1 + m2); // velocity of center-of-mass
		Accelerate((rv2 - center) * (m2 * coefficient) / m1 - (rv1 - center));
		other->Accelerate((rv1 - center) * (m1 * coefficient) / m2 - (rv2 - center));
	}
	GBDistance totalr = radius + other->radius;
	GBDistance overlap = totalr - dist;
	if ( overlap > kMaxOverlap ) {
		GBVector away1 = - cc / dist * m2 / (m1 + m2);
		GBVector away2 = cc / dist * m1 / (m1 + m2);
		MoveBy(away1 * kCollisionRepulsionSpeed);
		other->MoveBy(away2 * kCollisionRepulsionSpeed);
		Accelerate(away1 * kCollisionRepulsion);
		other->Accelerate(away2 * kCollisionRepulsion);
	}
}

void GBObject::Think(GBWorld *) {}

void GBObject::Move() {
	position += velocity;
}

void GBObject::Act(GBWorld *) {}

void GBObject::CollideWithWall() {}

void GBObject::CollideWith(GBObject *) {}

void GBObject::CollectStatistics(GBWorld *) const {}

GBObjectClass GBObject::Class() const {
	return ocDecoration;
}

GBSide * GBObject::Owner() const {
	return nil;
}

GBEnergy GBObject::Energy() const {
	return 0;
}

GBNumber GBObject::Interest() const {
	return 0;
}

std::string GBObject::Description() const {
	return "Generic object";
}

std::string GBObject::Details() const {
	return "";
}

const GBColor GBObject::Color() const {
	return GBColor(1);
}

void GBObject::Draw(GBGraphics & g, const GBRect & where, bool /*detailed*/) const {
	g.DrawOpenOval(where, Color());
}

void GBObject::DrawMini(GBGraphics & g, const GBRect & where) const {
	if ( where.Width() < kMaxSquareMiniSize )
		g.DrawSolidRect(where, Color());
	else
		g.DrawSolidOval(where, Color());
}

